rdmd.d

```diff
@@ -839,6 +835,102 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
     std.zlib;
 ";
 
+/**
+    Joins together the code provided via an `--eval`
+    or `--loop` flag, ensuring a trailing `;` is
+    added if not already provided by the user
+
+    Params:
+        eval = array of strings generated by the
+               `--eval` or `--loop` rdmd flags
+
+    Returns:
+        string of code to be evaluated, corresponding
+        to the inner code of either the program or
+        the loop
+*/
+string innerEvalCode(string[] eval)
+{
+    import std.string: join, stripRight;
+    string code = eval.join("\n");
+    if (stripRight(eval[$ - 1])[$ - 1] != ';')
+        code ~= ';';
+    return code;
+}
+
+unittest
+{
+    assert(innerEvalCode([`writeln("Hello!")`]) == `writeln("Hello!");`);
+    assert(innerEvalCode([`writeln("Hello!");`]) == `writeln("Hello!");`);
+
+    // test with trailing whitespace
+    assert(innerEvalCode([`writeln("Hello!")  `]) == `writeln("Hello!")  ;`);
+    assert(innerEvalCode([`writeln("Hello!");  `]) == `writeln("Hello!");  `);
+
+    // test with multiple entries
+    assert(innerEvalCode([`writeln("Hello!");  `, `writeln("You!")  `])
+           == "writeln(\"Hello!\");  \nwriteln(\"You!\")  ;");
+    assert(innerEvalCode([`writeln("Hello!");  `, `writeln("You!"); `])
+           == "writeln(\"Hello!\");  \nwriteln(\"You!\"); ");
+}
+
+/**
+    Formats the code provided via `--eval` or `--loop`
+    flags into a string of complete program code that
+    can be written to a file and then compiled
+
+    Params:
+        eval = array of strings generated by the
+               `--eval` or `--loop` rdmd flags
+        loop = set to `true` if this code comes
+               from a `--loop` flag, `false` (the
+               default) otherwise
+
+    Returns:
+        string of code to be evaluated, corresponding
+        to the inner code of either the program or
+        the loop
+*/
+string makeEvalCode(string[] eval, bool loop = false)
+{
+    import std.format : format;
+    immutable codeFormat = importWorld
+        ~ "void main(char[][] args) {%s%s\n%s}";
+
+    immutable innerCodeOpening =
+        loop ? " foreach (line; std.stdio.stdin.byLine()) {\n"
```

Comment:
Idiomatic D would just use `.byLine` - you could use this opportunity to change this bit.